third/nmh/man/ap.man
.\"
.\" %nmhwarning%
.\" $Id: ap.man,v 1.1.1.1 1999-02-07 18:14:19 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH AP %manext8% MH.6.8 [%nmhversion%]
.SH NAME
ap \- parse addresses 822-style
.SH SYNOPSIS
.in +.5i
.ti -.5i
%libdir%/ap
\%[\-form\ formatfile]
.br
\%[\-format\ string]
\%[\-normalize] \%[\-nonormalize]
.br
\%[\-width\ columns]
addrs\ ...
.br
\%[\-version]
\%[\-help] 
.in -.5i
.SH DESCRIPTION
\fIAp\fR is a program that parses addresses according to the ARPA
Internet standard.  It also understands many non\-standard formats.
It is useful for seeing how \fInmh\fR will interpret an address.

The \fIap\fR program treats each argument as one or more addresses, and
prints those addresses out in the official 822\-format.  Hence, it is
usually best to enclose each argument in double\-quotes for the shell.

To override the output format used by \fIap\fR, the `\-format\ string' or
`\-format\ file' switches are used.  This permits individual fields of
the address to be extracted with ease.  The string is simply a format
string, and the file is simply a format file.  See \fImh\-format\fR\0(5)
for the details.

In addition to the standard escapes,
\fIap\fR also recognizes the following additional escape:
.sp 1
.nf
.ta \w'Escape  'u +\w'Returns  'u
\fIEscape\fR	\fIReturns\fR	\fIDescription\fR
error	string	A diagnostic if the parse failed
.re
.fi

If the `\-normalize' switch is given, \fIap\fR will try to track down
the official hostname of the address.

Here is the default format string used by \fIap\fR:

.ti +.5i
%<{error}%{error}: %{text}%|%(putstr(proper{text}))%>

which says that if an error was detected, print the error, a `:', and
the address in error.  Otherwise, output the 822\-proper format of
the address.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
^%etcdir%/mts.conf~^nmh mts configuration file
.Pr
None
.Sa
dp(8),
.br
\fIStandard for the Format of ARPA Internet Text Messages\fR (RFC\-822)
.De
`\-format' defaults as described above
.Ds
`\-normalize'
.Ds
`\-width' defaults to the width of the terminal
.Co
None
.Bu
The argument to the `\-format' switch must be interpreted as a single token
by the shell that invokes \fIap\fR.
Therefore,
one must usually place the argument to this switch inside double\-quotes.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/ap.c

/*
 * ap.c -- parse addresses 822-style
 *
 * $Id: ap.c,v 1.1.1.1 1999-02-07 18:14:12 danw Exp $
 */

#include <h/mh.h>
#include <h/addrsbr.h>
#include <h/fmt_scan.h>

#define	NADDRS	100

#define	WIDTH	78
#define	WBUFSIZ	BUFSIZ

#define	FORMAT	"%<{error}%{error}: %{text}%|%(putstr(proper{text}))%>"

static struct swit switches[] = {
#define	FORMSW	0
    { "form formatfile", 0 },
#define	FMTSW	1
    { "format string", 5 },
#define	NORMSW	2
    { "normalize", 0 },
#define	NNORMSW	3
    { "nonormalize", 0 },
#define	WIDTHSW	4
    { "width columns", 0 },
#define VERSIONSW 5
    { "version", 0 },
#define	HELPSW	6
    { "help", 4 },
    { NULL, 0 }
};

static struct format *fmt;

static int dat[5];

/*
 * prototypes
 */
int sc_width (void);  /* from termsbr.c */

/*
 * static prototypes
 */
static int process (char *, int, int);


int
main (int argc, char **argv)
{
    int addrp = 0, normalize = AD_HOST;
    int width = 0, status = 0;
    char *cp, *form = NULL, *format = NULL, *nfs;
    char buf[BUFSIZ], **argp;
    char **arguments, *addrs[NADDRS];

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);

		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] addrs ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version (invo_name);
		    done (1);

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW: 
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    width = atoi (cp);
		    continue;

		case NORMSW: 
		    normalize = AD_HOST;
		    continue;
		case NNORMSW: 
		    normalize = AD_NHST;
		    continue;
	    }
	}
	if (addrp > NADDRS)
	    adios (NULL, "more than %d addresses", NADDRS);
	else
	    addrs[addrp++] = cp;
    }
    addrs[addrp] = NULL;

    if (addrp == 0)
	adios (NULL, "usage: %s [switches] addrs ...", invo_name);

    /* get new format string */
    nfs = new_fs (form, format, FORMAT);

    if (width == 0) {
	if ((width = sc_width ()) < WIDTH / 2)
	    width = WIDTH / 2;
	width -= 2;
    }
    if (width > WBUFSIZ)
	width = WBUFSIZ;
    fmt_norm = normalize;
    fmt_compile (nfs, &fmt);

    dat[0] = 0;
    dat[1] = 0;
    dat[2] = 0;
    dat[3] = width;
    dat[4] = 0;

    for (addrp = 0; addrs[addrp]; addrp++)
	status += process (addrs[addrp], width, normalize);

    done (status);
}

struct pqpair {
    char *pq_text;
    char *pq_error;
    struct pqpair *pq_next;
};


static int
process (char *arg, int length, int norm)
{
    int	status = 0;
    register char *cp;
    char buffer[WBUFSIZ + 1], error[BUFSIZ];
    register struct comp *cptr;
    register struct pqpair *p, *q;
    struct pqpair pq;
    register struct mailname *mp;

    (q = &pq)->pq_next = NULL;
    while ((cp = getname (arg))) {
	if ((p = (struct pqpair *) calloc ((size_t) 1, sizeof(*p))) == NULL)
	    adios (NULL, "unable to allocate pqpair memory");
	if ((mp = getm (cp, NULL, 0, norm, error)) == NULL) {
	    p->pq_text = getcpy (cp);
	    p->pq_error = getcpy (error);
	    status++;
	}
	else {
	    p->pq_text = getcpy (mp->m_text);
	    mnfree (mp);
	}
	q = (q->pq_next = p);
    }

    for (p = pq.pq_next; p; p = q) {
	FINDCOMP (cptr, "text");
	if (cptr)
	    cptr->c_text = p->pq_text;
	FINDCOMP (cptr, "error");
	if (cptr)
	    cptr->c_text = p->pq_error;

	fmt_scan (fmt, buffer, length, dat);
	fputs (buffer, stdout);

	free (p->pq_text);
	if (p->pq_error)
	    free (p->pq_error);
	q = p->pq_next;
	free ((char *) p);
    }

    return status;
}
